tool-openssl/req.cc

```diff
@@ -358,128 +457,314 @@ static int make_certificate_request(X509_REQ *req, EVP_PKEY *pkey,
+  int idx = X509_get_ext_by_OBJ(cert, keyid_ext_obj, -1);
+  if (idx >= 0) {
+    return 1;  // Extension found
+  }
+
+  return !add_if_missing || X509_add_ext(cert, new_keyid_ext.get(), -1);
 }
 
-// Function to add extensions to a certificate
-static bool add_cert_extensions(X509 *cert) {
-  const char *config =
-      "[v3_ca]\n"
-      "subjectKeyIdentifier=hash\n"
-      "authorityKeyIdentifier=keyid:always,issuer:always\n"
-      "basicConstraints=critical,CA:true\n";
-
-  // Create a BIO for the config
-  bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(config, -1));
-  if (!bio) {
-    fprintf(stderr, "Failed to create memory BIO\n");
+static bool LoadConfig(const std::string &config_path,
+                       bssl::UniquePtr<CONF> &conf) {
+  bssl::UniquePtr<BIO> conf_bio(BIO_new_file(config_path.c_str(), "r"));
+  if (!conf_bio) {
+    fprintf(stderr, "Error: unable to load extension file from '%s'\n",
+            config_path.c_str());
     return false;
   }
 
-  bssl::UniquePtr<CONF> conf(NCONF_new(NULL));
+  conf.reset(NCONF_new(NULL));
   if (!conf) {
-    fprintf(stderr, "Failed to create CONF structure\n");
+    fprintf(stderr, "Error: Failed to create extension config\n");
     return false;
   }
 
-  if (NCONF_load_bio(conf.get(), bio.get(), NULL) <= 0) {
-    fprintf(stderr, "Failed to load config from BIO\n");
+  if (NCONF_load_bio(conf.get(), conf_bio.get(), NULL) <= 0) {
+    fprintf(stderr, "Error: Failed to load config from BIO\n");
     return false;
   }
 
+  return true;
+}
+
+static bool AddCertExtensions(X509 *cert, CONF *req_conf,
+                              std::string &ext_section) {
+  // Set up X509V3 context for certificate
+  bool result = false;
+  X509V3_CTX ext_ctx;
+  X509V3_set_ctx(&ext_ctx, cert, cert, NULL, NULL,
+                 X509V3_CTX_REPLACE);  // self-signed
+
+  if (req_conf != NULL) {
+    X509V3_set_nconf(&ext_ctx, req_conf);
+
+    // Add extensions from config to the certificate
+    result = X509V3_EXT_add_nconf(req_conf, &ext_ctx, ext_section.c_str(),
+                                  cert) != 0;
+
+    /* Prevent X509_V_ERR_MISSING_SUBJECT_KEY_IDENTIFIER */
+    if (!AdaptKeyIDExtension(cert, &ext_ctx, "subjectKeyIdentifier", "hash",
+                             1)) {
+      fprintf(stderr, "Error: Failed to handle subject key identifier\n");
+      return false;
+    }
+    /* Prevent X509_V_ERR_MISSING_AUTHORITY_KEY_IDENTIFIER */
+    if (!AdaptKeyIDExtension(cert, &ext_ctx, "authorityKeyIdentifier",
+                             "keyid, issuer", 0)) {
+      fprintf(stderr, "Error: Failed to handle authority key identifier\n");
+      return false;
+    }
+  } else {
+    const char *default_config =
+        "[v3_ca]\n"
+        "subjectKeyIdentifier=hash\n"
+        "authorityKeyIdentifier=keyid:always,issuer:always\n"
+        "basicConstraints=critical,CA:true\n";
+
+    // Create a BIO for the config
+    bssl::UniquePtr<BIO> bio(BIO_new_mem_buf(default_config, -1));
```

Comment:
do we have an upper bound on the buffer size here? negative length is not encouraged for new usages

https://github.com/aws/aws-lc/blob/981c79b69b930e17105b8ce1f20f7d81c97fe79c/include/openssl/bio.h#L461-L463

Reply:
I updated this to provide the string length.